### PR TITLE
Akcore 40: Acknowledge implementation in the SharePartitionManager

### PR DESCRIPTION
**Pull request by chirag-wadhwa5**
This PR contains the implementation of acknowledge function inside SharePartitionManager.java. This function is called from handleShareAcknowledgeRequest handler in KafkaApis.scala. 

Reference - [AKCORE-40](https://confluentinc.atlassian.net/jira/software/c/projects/AKCORE/issues/AKCORE-40)

[AKCORE-40]: https://confluentinc.atlassian.net/browse/AKCORE-40?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core/src/main/java/kafka/server/SharePartitionManager.java [hunk 1]
```diff
@@ -125,10 +126,46 @@ public CompletableFuture<Map<TopicIdPartition, ShareFetchResponseData.PartitionD
...
     public CompletableFuture<Map<TopicIdPartition, ShareAcknowledgeResponseData.PartitionData>> acknowledge(
+            String memberId,
             String groupId,
             Map<TopicIdPartition, ShareAcknowledgeRequestData.AcknowledgePartition> acknowledgeTopics
     ) {
-        throw new UnsupportedOperationException("Not implemented yet");
+        log.debug("Acknowledge request for topicIdPartitions: {} with groupId: {}",
+                acknowledgeTopics.keySet(), groupId);
+        Map<TopicIdPartition, CompletableFuture<Errors>> futures = new HashMap<>();
+        synchronized (this) {
```
Comment (AndrewJSchofield) [inline core/src/main/java/kafka/server/SharePartitionManager.java:136]: This is currently synchronized across the entire broker, which seems excessive.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]
Thanks for the PR.

File: core/src/main/java/kafka/server/SharePartitionManager.java (hunk 1 above)
Comment (chirag-wadhwa5) [reply core/src/main/java/kafka/server/SharePartitionManager.java]: Thanks for the review. I have made the changes in the latest commit

File: clients/src/main/java/org/apache/kafka/clients/consumer/AcknowledgeType.java [hunk 2]
```diff
@@ -25,4 +25,17 @@ public enum AcknowledgeType {
     public String toString() {
         return super.toString().toLowerCase(Locale.ROOT);
     }
+
+    public static AcknowledgeType acknowledgeTypeFromNumber(int a) {
+        switch (a) {
```
Comment (apoorvmittal10) [inline clients/src/main/java/org/apache/kafka/clients/consumer/AcknowledgeType.java:30]: Hmmm, might be better to associate with an id in constructor and change method name to `forId` as that's being don in other enums like CompressionType.java.

File: core/src/main/java/kafka/server/SharePartition.java [hunk 3]
```diff
@@ -121,7 +120,7 @@ public void update(String memberId, FetchPartitionData fetchPartitionData) {
         }
     }
 
-    public CompletableFuture<Optional<Throwable>> acknowledge(String memberId, List<AcknowledgementBatch> acknowledgementBatch) {
+    public CompletableFuture<Errors> acknowledge(String memberId, List<AcknowledgementBatch> acknowledgementBatch) {
```
Comment (apoorvmittal10) [inline core/src/main/java/kafka/server/SharePartition.java:123]: Any reason for this change?

File: core/src/main/java/kafka/server/SharePartitionManager.java [hunk 4]
```diff
@@ -125,10 +126,46 @@ public CompletableFuture<Map<TopicIdPartition, ShareFetchResponseData.PartitionD
...
             Map<TopicIdPartition, ShareAcknowledgeRequestData.AcknowledgePartition> acknowledgeTopics
     ) {
-        throw new UnsupportedOperationException("Not implemented yet");
+        log.debug("Acknowledge request for topicIdPartitions: {} with groupId: {}",
+                acknowledgeTopics.keySet(), groupId);
+        Map<TopicIdPartition, CompletableFuture<Errors>> futures = new HashMap<>();
+        acknowledgeTopics.forEach((topicIdPartition, acknowledgePartitionData) -> {
+            SharePartition sharePartition = partitionCacheMap.get(sharePartitionKey(groupId, topicIdPartition));
+            if (sharePartition != null) {
+                synchronized (sharePartition) {
```
Comment (apoorvmittal10) [inline core/src/main/java/kafka/server/SharePartitionManager.java:139]: May be too early to think about share partition synchronization as parallel requests for acknowledge should be thread safe by share partition itself, may be, we can keep this code for now but I don't think we ll need synchronization here.

File: core/src/main/java/kafka/server/SharePartitionManager.java [hunk 5]
```diff
@@ -125,10 +126,46 @@ public CompletableFuture<Map<TopicIdPartition, ShareFetchResponseData.PartitionD
...
+                    List<SharePartition.AcknowledgementBatch> acknowledgementBatches =
+                            acknowledgePartitionData.acknowledgementBatches()
+                            .stream()
+                            .map(batch -> new SharePartition.AcknowledgementBatch(
+                                    batch.startOffset(),
+                                    batch.lastOffset(),
+                                    batch.gapOffsets(),
+                                    AcknowledgeType.acknowledgeTypeFromNumber(batch.acknowledgeType())))
```
Comment (apoorvmittal10) [inline core/src/main/java/kafka/server/SharePartitionManager.java:147]: May be you want to handle exception occurring from this call.

File: core/src/main/java/kafka/server/SharePartition.java (hunk 3 above)
Comment (chirag-wadhwa5) [reply core/src/main/java/kafka/server/SharePartition.java:123]: We are using Errors everywhere, so I thought returning an object of Errors class would make more sense.

**Review by apoorvmittal10 — APPROVED** [review]

**Pull request merged**